common/changes/@rushstack/heft/typescript-5-compatibility_2023-02-28-00-08.json

```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/heft",
+      "comment": "Remove monkey-patching of TypeScript for compatibility with 5.0. Refactors how the multi-emit logic works.",
```

Comment:
@dmichon-msft just to confirm, the monkey patch is not "removed" -- the code is still used to support older versions of TypeScript, right?

Reply:
I'm not @dmichon-msft but it's fully removed; I'm pretty sure that the way this PR works should work for old versions of TS just fine.

Reply:
It works in both 4.7.4 and typescript@next. It doesn't rely on any net-new APIs from TypeScript to implement the functionality, simply a change to what we override (properties on Program instances instead of global properties of the `ts` namespace object). One important result is that this interaction with the TypeScript compiler now completely avoids any global state pollution.

The two methods `emit` and `getCompilerOptions` have not changed their signatures in the time that Heft has existed, so it should be fully backwards compatible to the earliest version that works with Heft.

Reply:
> The two methods `emit` and `getCompilerOptions` have not changed their signatures in the time that Heft has existed, so it should be fully backwards compatible to the earliest version that works with Heft.

Cool. If it's not difficult, please test an old TypeScript version to make sure.

Reply:
Validated that code builds with multi-emit using this change on TypeScript 2.9.2 (lowest version we officially support), although there is a gap from an earlier PR that this error code `Element_implicitly_has_an_any_type_because_expression_of_type_0_can_t_be_used_to_index_type_1` does not exist in earlier versions, so the error classifier breaks reporting.

Reply:
Fixed the reporter.